runtime-transaction/src/runtime_transaction.rs

```diff
@@ -0,0 +1,164 @@
+///
+/// New Error:
+/// If Getter of account-based metadata is called while in Initial state,
+/// a new Error shall return to indicate RuntimeTransaction is not
+/// furnished with necessary data yet.
+///
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+pub enum RuntimeMessage {
+    #[default]
+    Uninitialized,
+    StaticallyLoaded(SanitizedVersionedMessage),
+    DynamicallyLoaded(SanitizedMessage),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RuntimeTransaction {
+    // sanitized signatures
+    signatures: Vec<Signature>,
+
+    // sanitized message that can be in static/dynamic state
+    message: RuntimeMessage,
+
+    // transaction meta is a collection of fields, it is updated
+    // during message state transition
+    meta: TransactionMeta,
+}
+
+impl Default for RuntimeTransaction {
+    fn default() -> Self {
+        Self {
+            signatures: vec![],
+            message: RuntimeMessage::Uninitialized,
+            meta: TransactionMeta::default(),
+        }
+    }
+}
+
+impl RuntimeTransaction {
+    pub fn try_from(
+        sanitized_versioned_tx: SanitizedVersionedTransaction,
+        message_hash: Option<Hash>,
+        is_simple_vote_tx: Option<bool>,
+    ) -> Result<Self> {
+        let mut tx = RuntimeTransaction::default();
+        tx.to_statically_loaded(sanitized_versioned_tx, message_hash, is_simple_vote_tx)?;
+        Ok(tx)
+    }
+
+    // transit from Uninitialized to StaticallyLoaded state, where static metadata are available
+    pub fn to_statically_loaded(
+        &mut self,
+        sanitized_versioned_tx: SanitizedVersionedTransaction,
+        message_hash: Option<Hash>,
+        is_simple_vote_tx: Option<bool>,
+    ) -> Result<()> {
+        if let RuntimeMessage::Uninitialized = self.message {
+            self.load_static_metadata(&sanitized_versioned_tx, message_hash, is_simple_vote_tx)?;
+            self.signatures = sanitized_versioned_tx.signatures;
+            self.message = RuntimeMessage::StaticallyLoaded(sanitized_versioned_tx.message);
+        } else {
+            todo!();
+        }
+        Ok(())
+    }
+    
+    pub fn to_dynamically_loaded(
+        &mut self,
+        address_loader: impl AddressLoader,
+    ) -> Result<()> {
+        let message = std::mem::take(&mut self.message);
+        if let RuntimeMessage::StaticallyLoaded(sanitized_versioned_message) = message {
+            // load dynamic address accounts
+            let sanitized_message = SanitizedMessage::try_new(sanitized_versioned_message, address_loader)?;
+            let _ = std::mem::replace(&mut self.message, RuntimeMessage::DynamicallyLoaded(sanitized_message));
+            self.load_dynamic_metadata()?;
+        } else {
+            todo!();
+        }
+        Ok(())
+    }
+
+    // Getter
+    pub fn get_message_hash(&self) -> &Hash {
+        if let RuntimeMessage::Uninitialized = self.message {
+            panic!("Runtime transaction is uninitialized");
+        }
```

Comment:
panic or return Option when getter is called in wrong state? I pick panicking to fail hard, but mostly to keep interface simple.